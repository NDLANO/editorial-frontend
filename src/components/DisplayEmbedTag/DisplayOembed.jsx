/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { injectT } from 'ndla-i18n';
import EditorErrorMessage from '../SlateEditor/EditorErrorMessage';
import { fetchOembed } from '../../util/apiHelpers';

export const getIframeSrcFromHtmlString = html => {
  const el = document.createElement('html');
  el.innerHTML = html;
  const iframe = el.getElementsByTagName('iframe')[0];
  return iframe.getAttribute('src');
};

export class DisplayOembed extends React.Component {
  constructor(props) {
    super(props);
    this.state = {};
    this.handleResizeMessage = this.handleResizeMessage.bind(this);
  }

  async componentWillMount() {
    this.enableIframeResizing();
    try {
      const data = await fetchOembed(this.props.url);
      const src = getIframeSrcFromHtmlString(data.html);
      if (src) {
        this.setState({ title: data.title, src });
      } else {
        this.setState({ error: true });
      }
    } catch (e) {
      this.setState({ error: true });
    }
  }

  componentWillUnmount() {
    this.disableIframeResizing();
  }

  enableIframeResizing() {
    window.addEventListener('message', this.handleResizeMessage);
  }

  disableIframeResizing() {
    window.removeEventListener('message', this.handleResizeMessage);
  }

  handleResizeMessage(evt) {
    const iframe = this.iframe;

    if (
      !iframe ||
      iframe.contentWindow !== evt.source ||
      evt.data.context !== 'h5p'
    ) {
      return;
    }

    /* Needed to enforce content to stay within iframe on Safari iOS */
    iframe.setAttribute('scrolling', 'no');

    const newHeight = evt.data.scrollHeight ? evt.data.scrollHeight + 35 : 800;

    iframe.style.height = `${newHeight}px`;
  }

  render() {
    const { title, src, error } = this.state;

    if (error) {
      return (
        <EditorErrorMessage msg={this.props.t('displayOembed.errorMessage')} />
      );
    }

    return (
      <iframe
        ref={iframe => {
          this.iframe = iframe;
        }}
        src={src}
        title={title}
        frameBorder="0"
      />
    );
  }
}

DisplayOembed.propTypes = {
  url: PropTypes.string.isRequired,
};

export default injectT(DisplayOembed);
