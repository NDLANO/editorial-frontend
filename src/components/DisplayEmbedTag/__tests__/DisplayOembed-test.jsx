/**
 * Copyright (c) 2017-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import nock from 'nock';
import React from 'react';
import renderer from 'react-test-renderer';
import { DisplayOembed, getIframeSrcFromHtmlString } from '../DisplayOembed';

test('getIframeSrcFromHtmlString returns src attribute', () => {
  const src = getIframeSrcFromHtmlString(
    `<div>
      <iframe
        width="800"
        height="600"
        src="https://h5p-test.ndla.no/resource/3ab6850d">
      </iframe>
      <script src="https://h5p.org/h5p-resizer.js"></script>
    </div>`,
  );
  expect(src).toBe('https://h5p-test.ndla.no/resource/3ab6850d');
});

test('getIframeSrcFromHtmlString returns null id src not found', () => {
  const src = getIframeSrcFromHtmlString(
    `<div>
      <iframe
        width="800"
        height="600"
      </iframe>
      <script src="https://h5p.org/h5p-resizer.js"></script>
    </div>`,
  );
  expect(src).toBe(null);
});

test('DisplayOembed renderers correctly', async () => {
  nock('https://ndla.no/oembed').get('').reply(200, {
    title: 'unit test',
    type: 'rich',
    html: '<iframe src="iframe.html">',
  });

  const component = renderer.create(
    <DisplayOembed url="https://ndla.no/oembed" t={() => ''} />,
  );

  expect(component.toJSON()).toMatchSnapshot();

  return new Promise(resolve => {
    setTimeout(() => {
      expect(component.toJSON()).toMatchSnapshot();
      resolve();
    }, 100);
  });
});
