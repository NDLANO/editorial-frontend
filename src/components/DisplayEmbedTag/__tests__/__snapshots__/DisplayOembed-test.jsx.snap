// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisplayOembed renderers correctly 1`] = `
<iframe
  frameBorder="0"
  src={undefined}
  title={undefined}
/>
`;

exports[`DisplayOembed renderers correctly 2`] = `
<iframe
  frameBorder="0"
  src="iframe.html"
  title="unit test"
/>
`;
